numpy/core/src/multiarray/methods.c

```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
     return c;
 }
 
+#define NPY_DLPACK_CAPSULE_NAME "NumPy DLPack Wrapper"
+
+static void array_dlpack_capsule_deleter(PyObject *self)
+{
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_CAPSULE_NAME);
+    managed->deleter(managed);
+}
+
+static void array_dlpack_deleter(DLManagedTensor *self)
+{
+    PyArrayObject *array = (PyArrayObject *)self->manager_ctx;
+    free(self->dl_tensor.shape);
+    free(self->dl_tensor.strides);
+    free(self);
+
+    PyArray_XDECREF(array);
+}
+
+static PyObject *
+array_dlpack(PyArrayObject *self,
+        PyObject *const *args, Py_ssize_t len_args, PyObject *kwnames)
+{
+    PyObject *stream = Py_None;
+    NPY_PREPARE_ARGPARSER;
+    if (npy_parse_arguments("__dlpack__", args, len_args, kwnames,
+                             "$stream", NULL, &stream,
+                             NULL, NULL, NULL))
+    {
+        return NULL;
+    }
+
+    if (stream != Py_None)
+    {
+        PyErr_SetString(PyExc_RuntimeError, "NumPy only supports stream=None.");
+        return NULL;
+    }
+
+    npy_intp itemsize = PyArray_ITEMSIZE(self);
+    int ndim = PyArray_NDIM(self);
+    npy_intp *strides = PyArray_STRIDES(self);
+    npy_intp *shape = PyArray_SHAPE(self);
+
+    for (int i = 0; i < ndim; ++i)
+    {
+        if (strides[i] % itemsize != 0) {
+            PyErr_SetString(PyExc_RuntimeError,
+                            "DLPack only supports strides which are a multiple of itemsize.");
+            return NULL;
+        }
+    }
+
+    DLDataType managed_dtype;
+    PyArray_Descr *dtype = PyArray_DESCR(self);
+    
+    managed_dtype.bits = 8 * itemsize;
+    managed_dtype.lanes = 1;
+    if (PyDataType_ISSIGNED(dtype))
+    {
+        managed_dtype.code = kDLInt;
+    }
+    else if (PyDataType_ISUNSIGNED(dtype))
+    {
+        managed_dtype.code = kDLUInt;
+    }
+    else if (PyDataType_ISFLOAT(dtype))
+    {
+        managed_dtype.code = kDLFloat;
+    }
+    else if (PyDataType_ISCOMPLEX(dtype))
+    {
+        managed_dtype.code = kDLComplex;
+    }
+    else
+    {
+        PyErr_SetString(PyExc_TypeError,
+                        "DLPack only supports signed/unsigned integers, float and complex dtypes.");
```

Comment:
Just to verify: DLPack doesn't support boolean dtypes?

Reply:
It could in theory, but it doesn't as of version 0.5.0.